# 6.- MODELO 2: SVM

SVMs pueden usarse como algoritmo de clasificación y regresión, en este caso es una clasifocación lo que estamos buscando, el cliente hará o no una reclamación económica.

Dado un conjunto de muestras de entrenamiento, podemos etiquetar las clases y entrenar con ellas el modelo SVM, que nos servirá para predecir la clase de una nueva muestra.

Una SVM representa los puntos de una muestra en el espacio, y separa las clases en dos espacios lo más amplios posibles, mediante un hiperplano de separación que se define por los 2 puntos de las 2 clases más cercanas, y que se llama vector soporte. Cuando utilizamos una nueva muestra sobre este modelo, podemos saber a qué clase pertenece, según si queda clasificada en un parte u en otra del hiperplano. Funcionan con un principio de discriminación, en el que buscamos un hiperplano para separar (clasificar) nuestros datos, que esté métricamente más separado de las observaciones. En problemas de varias dimensiones (variables), la aproximación común es generar una nueva dimensión (kernel), con la que podamos encontrar el hiperplano más facilmente.

In [25]:
#Importación de bibliotecas
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [26]:
# Importo los datos
df = pd.read_csv("df_finalVF.csv",  sep = ";")

In [27]:
df.head()

Unnamed: 0.1  Unnamed: 0  DRIVING_EXPERIENCE  INCOME  VEHICLE_OWNERSHIP  \
0             0           0                   0     2.0                1.0   
1             1           1                   0     0.0                0.0   
2             2           2                   0     1.0                1.0   
3             3           3                   0     1.0                1.0   
4             4           4                   1     1.0                1.0   

   MARRIED  CHILDREN  ANNUAL_MILEAGE  SPEEDING_VIOLATIONS  DUIS  \
0      0.0       1.0         12000.0                    0     0   
1      0.0       0.0         16000.0                    0     0   
2      0.0       0.0         11000.0                    0     0   
3      0.0       1.0         11000.0                    0     0   
4      0.0       0.0         12000.0                    2     0   

   PAST_ACCIDENTS  OUTCOME  GENDER_female  RACE_majority  \
0               0      0.0              1              1   
1               0      1.0              0              1   
2               0      0.0              1              1   
3               0      0.0              0              1   
4               1      1.0              0              1   

   VEHICLE_YEAR_before 2015  VEHICLE_TYPE_sedan  
0                         0                   1  
1                         1                   1  
2                         1                   1  
3                         1                   1  
4                         1                   1

In [28]:
df = df.drop(labels='ANNUAL_MILEAGE', axis=1)

In [29]:
# Separo la variable target
X = df.drop('OUTCOME', axis=1)
y = df['OUTCOME']


In [30]:
# Separo Train / Test: Dejo un 80% del dataset para train y 20% para test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20)

El SVM se lleva a cabo con un aumento de la dimensión (un hiperplano), para esto se usa el " truco kernel". Existen multitud de kernels distintos, en este modelo vamos a probar tres: 1) Kernel Polinómico, 2) Kernel Gaussiano, 3) Kernel Sigmoideo, para ver cual nos da mejor resultado a la hora de predecir una clasificación de la variable OUTCOME con este dataset



### 1) Kernel Polinomico

In [56]:
from sklearn.svm import SVC
svclassifier = SVC(kernel="poly", degree=3)
svclassifier.fit(X_train, y_train)

SVC(kernel='poly')

In [57]:
# Las predicciones: aplicamos el modelo a Test a ver que resultados da
y_pred = svclassifier.predict(X_test)

In [58]:
# Evaluamos los resultados del algoritmo en Test
from sklearn.metrics import classification_report, confusion_matrix
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[1265    0]
 [ 544    0]]
              precision    recall  f1-score   support

         0.0       0.70      1.00      0.82      1265
         1.0       0.00      0.00      0.00       544

    accuracy                           0.70      1809
   macro avg       0.35      0.50      0.41      1809
weighted avg       0.49      0.70      0.58      1809



### B) Kernel Gaussiano

In [49]:
from sklearn.svm import SVC
svclassifier = SVC(kernel="rbf")
svclassifier.fit(X_train, y_train)

SVC()

In [50]:
# Las predicciones: aplicamos el modelo a Test a ver que resultados da
y_pred = svclassifier.predict(X_test)

In [51]:
# Evaluamos los resultados del algoritmo en Test
from sklearn.metrics import classification_report, confusion_matrix
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[1265    0]
 [ 544    0]]
              precision    recall  f1-score   support

         0.0       0.70      1.00      0.82      1265
         1.0       0.00      0.00      0.00       544

    accuracy                           0.70      1809
   macro avg       0.35      0.50      0.41      1809
weighted avg       0.49      0.70      0.58      1809



## C) Núcleo sigmoideo

In [53]:
from sklearn.svm import SVC
svclassifier = SVC(kernel="sigmoid")
svclassifier.fit(X_train, y_train)

SVC(kernel='sigmoid')

In [54]:
# Las predicciones: aplicamos el modelo a Test a ver que resultados da
y_pred = svclassifier.predict(X_test)

In [55]:
from sklearn.metrics import classification_report, confusion_matrix
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[865 400]
 [369 175]]
              precision    recall  f1-score   support

         0.0       0.70      0.68      0.69      1265
         1.0       0.30      0.32      0.31       544

    accuracy                           0.57      1809
   macro avg       0.50      0.50      0.50      1809
weighted avg       0.58      0.57      0.58      1809



**CONCLUSIONES**

Si comparamos el rendimiento de los diferentes tipos de kernel, podemos ver claramente que el sigmoide es el mejor. Esto se debe a que la función sigmoidea devuelve dos valores, 0 y 1, por lo que es más adecuada para problemas de clasificación binaria, como es nuestro caso. Sin embargo el nivel de falsos negativos que identifica es incluso mayor que los que acierta, y los falsos positivos son un 50% de los positivos que acierta. En conclusión su ratio de accuracy es solo del 57%.


El kernel gaussiano y el kernel polinomial tienen un desempeño desempeñó ligeramente mejor en cuanto el accuracy (70%), pero no realizan nunca una clasificación negativa (ni cierta ni falsa). Además en cuanto al kernel polinomico, se han efectuado pruebas con diferentes grados /eg. degree=1, 2, 3, 4), dando siempre el mismo resultado insatisfactorio.

La conclusión de este ejercicio es que el algorirmo SVM no realiza una buena labor a la hora de predecir una clasificación en este caso.
